### PR TITLE
fix(lib): add down broker to the broker list if it is brought back on…

### PR DESCRIPTION
**Pull request by outsinre**
If a partition is assigned to only 1 broker and unfortunately that broker is down, the broker is removed from the broker pool list.

This is expected. However, if the broker is brought back online, the code does not add back it to the broker pool list, and logs cannot be sent to Kafka unless a Kong restart.

This PR fixes this issue. It can now add the broker back if it is available again. There is no need to restart Kong.

Main changes.

1. Retry to get full broker list.
2. Add a new broker to form a cluster. This is quite useful as single node Kafka is rarely used in production.
3. Add tests to simulate broker2 down and up.
4.

### CODE REVIEW

File: lib/resty/kafka/sendbuffer.lua [hunk 1]
```diff
@@ -147,12 +151,26 @@ function _M.aggregator(self, client)
     local sendbroker = {}
     local brokers = {}
 
-    local i = 1
```
Comment (outsinre) [inline lib/resty/kafka/sendbuffer.lua:150]: This variable is declared but never used. So, remove it.

**Comment by outsinre** [comment]
Local tests work as expected.

image

**Comment by outsinre** [comment]
All local test passed:

image

File: lib/resty/kafka/sendbuffer.lua [hunk 2]
```diff
@@ -17,12 +19,14 @@ local MAX_REUSE = 10000
...
 local mt = { __index = _M }
 
-function _M.new(self, batch_num, batch_size)
+function _M.new(self, batch_num, batch_size, max_retry, retry_backoff)
     local sendbuffer = {
         topics = {},
         queue_num = 0,
         batch_num = batch_num * 2,
         batch_size = batch_size,
+        max_retry = max_retry or 3,
```
Comment (jschmid1) [inline lib/resty/kafka/sendbuffer.lua:28]: There are a lot of `max_retry or 3` invocations. Consider moving this to a more central place.
Same goes for `retry_backoff`

File: spec/producer_spec.lua [hunk 3]
```diff
@@ -231,19 +233,72 @@ describe("Test producers: ", function()
...
+
+    ngx_sleep(5)
+    offset, err = p:send("brokerdown", key, "afterstop message")
+    assert.is_nil(offset)
+    assert.is_same("not found broker; not found broker; not found broker", err)
+
+    ok = os.execute(string.format("docker compose -p dev %s broker2", "start"))
+    assert.is_truthy(ok)
+
+    ngx_sleep(15)  -- wait for rediscovery
```
Comment (jschmid1) [inline spec/producer_spec.lua:270]: Did you test for flakyness? Tests that are bound to timing tend to fail randomly.

File: dev/tokens/delegation-tokens.env [hunk 4]
```diff
@@ -0,0 +1,2 @@
+KYCaRy3-RYmtYfyU0DgWwg
```
[the remark this answers is not in the document]
Comment (outsinre) [reply dev/tokens/delegation-tokens.env:1]: I checked this file would be overwritten by the CI.

But we can remove the contents.

File: lib/resty/kafka/sendbuffer.lua (hunk 2 above)
Comment (outsinre) [reply lib/resty/kafka/sendbuffer.lua:28]: This retry here (`sendbuffer.lua`) is to retry the Kafka cluster nodes, while the retry in `producer.lua` is for retry to send message. They are for different purposes, but the value are both from Kong's config.

In `sendbuffer.lua` we cannot get the `retry` value from `producer.lua`.

File: dev/openresty/Dockerfile [hunk 5]
```diff
@@ -4,7 +4,16 @@ FROM ubuntu:20.04
 
 ENV DEBIAN_FRONTEND noninteractive
 RUN apt-get update && apt-get install -y --no-install-recommends apt-utils
-RUN apt-get install inotify-tools -y
+RUN apt-get -y install inotify-tools ca-certificates curl gnupg
```
[the remark this answers is not in the document]
Comment (outsinre) [reply dev/openresty/Dockerfile:7]: This just install a docker client CLI without the daemon part.

We use the CLI to control the daemon on the CI host (e.g. start/stop the newly added broker).

I tried other methods (e.g. customize `Makefile`), this is a more reliable way to simulate the broker down and up without mocking.

File: spec/producer_spec.lua (hunk 3 above)
Comment (outsinre) [reply spec/producer_spec.lua:270]: Tested. Most of the time, `1~5` seconds is enough, `15s` is a safeguard. I increase this value to make it robust.

Consider this repo is less updated than kong or kong-ee, I prefer not to increase the 'helpers.wait_until' utility.

**Review by jschmid1 — APPROVED** [review]

**Review by guanlan — APPROVED** [review]

**Pull request merged**